### PR TITLE
Introduce new HTTP BBEs

### PR DESCRIPTION
**Pull request by chamil321**

### CODE REVIEW

File: examples/http-default-resource/http_default_resource.bal [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import ballerina/http;
+
+// The `absolute resource path` can be omitted, then it defaults to `/`.
```
Comment (praneesha) [inline examples/http-default-resource/http_default_resource.bal:3]:
```suggestion
// The `absolute resource path` can be omitted. Then, it defaults to `/`.
```

File: examples/http-default-resource/http_default_resource.bal [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+service on new http:Listener(9090) {
+
+    // The `default` accessor name can be used to match with all methods including standard HTTP methods
+    // and custom methods. The rest param is used to represents the wildcard of `resource path` where any path
```
Comment (praneesha) [inline examples/http-default-resource/http_default_resource.bal:7]:
```suggestion
    // and custom methods. The rest param is used to represent the wildcard of the `resource path` in which any path
```

File: examples/http-default-resource/http_default_resource.bal [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    // segment will get dispatched to the resource in the absence of an exact path match.
+    resource function 'default [string... paths](http:Request req) returns json {
+        return {method: req.method, path: paths};
+    }
+
+    // The resource is confined to `GET` requests with path `/greeting`
```
Comment (praneesha) [inline examples/http-default-resource/http_default_resource.bal:13]:
```suggestion
    // The resource is confined to `GET` requests with the path `/greeting`.
```

File: examples/http-default-resource/http_default_resource.client.out [hunk 4]
```diff
@@ -0,0 +1,7 @@
+Run the cURL command below with unmatched path and method.
```
Comment (praneesha) [inline examples/http-default-resource/http_default_resource.client.out:1]:
```suggestion
Run the cURL command below with the unmatched path and method.
```

File: examples/http-default-resource/http_default_resource.client.out [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+curl "http://localhost:9090/foo/bar" -X POST
+{"method":"POST", "path":["foo", "bar"]}
+
+Run the cURL command below to invoke greeting resource.
```
Comment (praneesha) [inline examples/http-default-resource/http_default_resource.client.out:5]:
```suggestion
Run the cURL command below to invoke the `greeting` resource.
```

File: examples/http-default-resource/http_default_resource.description [hunk 6]
```diff
@@ -0,0 +1,4 @@
+// Ballerina provides rest params in the resource path and default resource method to help designing proxy services
```
Comment (praneesha) [inline examples/http-default-resource/http_default_resource.description:1]:
```suggestion
// Ballerina provides rest params in the resource path and the default resource method to help designing proxy services
```

File: examples/http-default-resource/tests/http_default_resource_test.bal [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+    json payload = {"method":"GET","path":["foo","bar"]};
+    var response = httpEndpoint->get("/foo/bar", targetType = json);
+    if (response is json) {
+        test:assertEquals(response, payload);
+    } else {
+        test:assertFail(msg = "Failed to call the endpoint");
+    }
+
+    string value = "Specific resource is invoked";
```
Comment (MadhukaHarith92) [inline examples/http-default-resource/tests/http_default_resource_test.bal:17]: Seems `value` is never reused. Shall we have `test:assertEquals(response, "Specific resource is invoked");` instead.

File: examples/http-path-matrix-param/http_path_matrix_param.bal [hunk 8]
```diff
@@ -23,7 +20,6 @@ service /sample on new http:Listener(9090) {
         // Create a JSON payload with the extracted values.
```
Comment (MadhukaHarith92) [inline examples/http-path-matrix-param/http_path_matrix_param.bal:20]:
```suggestion
        // Creates a JSON payload with the extracted values.
```

File: examples/http-path-matrix-param/http_path_matrix_param.bal [hunk 9]
```diff
@@ -2,12 +2,9 @@ import ballerina/http;
...
-    // The `PathParam` and `QueryParam` parameters extract values from the request URI.
-    // Path param is defined as a part of the resource path along with the type.
+    // The path param is defined as a part of the resource path along with the type and it is extracted from the
+    // request URI.
     resource function get path/[string foo](http:Request req) returns json {
-        // Get the [QueryParam](https://docs.central.ballerina.io/ballerina/http/latest/classes/Request#getQueryParamValue)
-        // value for a given parameter key.
-        var bar = req.getQueryParamValue("bar");
 
         // Get the [MatrixParams](https://docs.central.ballerina.io/ballerina/http/latest/classes/Request#getMatrixParams).
```
Comment (MadhukaHarith92) [inline examples/http-path-matrix-param/http_path_matrix_param.bal:9]:
```suggestion
        // Gets the [MatrixParams](https://docs.central.ballerina.io/ballerina/http/latest/classes/Request#getMatrixParams).
```

File: examples/http-path-matrix-param/http_path_matrix_param.bal [hunk 10]
```diff
@@ -23,7 +20,6 @@ service /sample on new http:Listener(9090) {
         // Create a JSON payload with the extracted values.
         json responseJson = {
             "pathParam": foo,
-            "queryParam": bar,
             "matrix": matrixJson
         };
         // Send a response with the JSON payload to the client.
```
Comment (MadhukaHarith92) [inline examples/http-path-matrix-param/http_path_matrix_param.bal:25]:
```suggestion
        // Sends a response with the JSON payload to the client.
```

File: examples/http-query-parameter/http_query_parameter.bal [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    // The multiple query param values also can be accommodate to an array.
+    resource function get detail(string[]? colour) returns json {
+        return { product_colour : colour};
+    }
+
+    // In addition to resource method argument, query params can be retrieved from the inbound request
+    resource function get part(http:Request req) returns json {
+        // Get the [first queryParam](https://docs.central.ballerina.io/ballerina/http/latest/classes/Request#getQueryParamValue)
```
Comment (MadhukaHarith92) [inline examples/http-query-parameter/http_query_parameter.bal:26]:
```suggestion
        // Gets the [first queryParam](https://docs.central.ballerina.io/ballerina/http/latest/classes/Request#getQueryParamValue)
```

File: examples/http-query-parameter/http_query_parameter.bal [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        // value for a given parameter key.
+        string? foo = req.getQueryParamValue("foo");
+
+        // Get [multiple queryParam](https://docs.central.ballerina.io/ballerina/http/latest/classes/Request#getQueryParamValues)
```
Comment (MadhukaHarith92) [inline examples/http-query-parameter/http_query_parameter.bal:30]:
```suggestion
        // Gets [multiple queryParam](https://docs.central.ballerina.io/ballerina/http/latest/classes/Request#getQueryParamValues)
```

File: examples/http-query-parameter/http_query_parameter.bal [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        // values for a given parameter key.
+        string[]? bar = req.getQueryParamValues("bar");
+
+        // Get [all the queryParam](https://docs.central.ballerina.io/ballerina/http/latest/classes/Request#getQueryParams)
```
Comment (MadhukaHarith92) [inline examples/http-query-parameter/http_query_parameter.bal:34]:
```suggestion
        // Gets [all the queryParam](https://docs.central.ballerina.io/ballerina/http/latest/classes/Request#getQueryParams)
```

File: examples/http-default-resource/tests/http_default_resource_test.bal [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    response = httpEndpoint->get("/greeting", targetType = string);
+    if (response is string) {
+        test:assertEquals(response, value);
```
Comment (chamil321) [inline examples/http-default-resource/tests/http_default_resource_test.bal:20]:
```suggestion
        test:assertEquals(response, "Specific resource is invoked");
```

File: examples/http-default-resource/tests/http_default_resource_test.bal (hunk 7 above)
Comment (chamil321) [inline examples/http-default-resource/tests/http_default_resource_test.bal:17]:
```suggestion
```

**Review by MadhukaHarith92 — APPROVED** [review]
LGTM

**Review by praneesha — APPROVED** [review]

**Pull request merged**